BlueprintUI/Tests/StackTests.swift

```diff
@@ -67,65 +67,285 @@ class StackTests: XCTestCase {
-        test(alignment: .trailing, layoutCrossSize: 200, elementCrossSize: 100, expectedOrigin: 100, expectedSize: 100)
-        test(alignment: .fill, layoutCrossSize: 200, elementCrossSize: 100, expectedOrigin: 0, expectedSize: 200)
-        test(alignment: .center, layoutCrossSize: 200, elementCrossSize: 100, expectedOrigin: 50, expectedSize: 100)
+        // In this test, 3 boxes are aligned to their top & bottom edges.
+        // The stack size matches the size of the content and needs no further alignment.
+        //
+        //          ┌ ─ ─ ─ ─ ─ ─ ┬──────────┐─ ─ ─ ─ ─ ─ ─
+        //                        │100       │             │
+        //          │             │          │
+        //                        │          │             │
+        //          │             │          │
+        //                        │          │             │
+        //        ─ ┼┌──────────┬ ┴──────────┘─┌──────────┬ ─ guide
+        //           │0         │              │0         ││
+        //          ││          │              │          │
+        //           │          │              │          ││
+        //          ││          │              │          │
+        //           │          │              │          ││
+        //  anchor ▶┴┴──────────┴──────────────┴──────────┴─◀
+        //
+        test(
+            stack: Row { row in
+                row.verticalAlignment = .bottom
+
+                // align to top edge
+                row.add(alignmentGuide: { _ in 0 }, child: TestElement())
+                // align to bottom edge
+                row.add(alignmentGuide: { d in d.height }, child: TestElement())
+                // align to top using another guide
+                row.add(alignmentGuide: { d in d[.top] }, child: TestElement())
+            },
+            layoutSize: CGSize(width: 300, height: 200),
+            expectedSize: CGSize(width: 300, height: 200),
+            expectedOrigins: [
+                100,
+                0,
+                100,
+            ]
+        )
+        // Same test on the other axis.
+        test(
+            stack: Column { column in
+                column.horizontalAlignment = .trailing
+
+                column.add(alignmentGuide: { _ in 0 }, child: TestElement())
+                column.add(alignmentGuide: { d in d.width }, child: TestElement())
+                column.add(alignmentGuide: { d in d[.leading] }, child: TestElement())
+            },
+            layoutSize: CGSize(width: 200, height: 300),
+            expectedSize: CGSize(width: 200, height: 300),
+            expectedOrigins: [
+                100,
+                0,
+                100,
+            ]
+        )
 
+        // In this test, 4 boxes are aligned relatively based on an arbitrary offset.
+        // The relatively aligned content is larger than the stack bounds,
+        // so the alignment is used to anchor the content to the center.
```

Comment:
I find it a bit odd that if the content is larger than the bounds; everything ends up being centered; eg it's weird one outlier element can cause others' positions to be changed. Does this match SwiftUI, or is it for another reason?

Reply:
> if the content is larger than the bounds; everything ends up being centered 

This is where the different layout models make things more complicated. In SwiftUI, the size of the stack is the size of its contents, so there is no concept of aligning the contents within the bounds — they are relatively aligned and then it's done. But in Blueprint, the size of the stack can be anything, so we still have to figure out where to place the contents after relatively aligning them.

So the stack's alignment serves two purposes here:
- first it's used during relative alignment to get the guide value on each child that doesn't have an explicit alignment guide
- then it's used to align the contents within the stack bounds

In this case the contents are centered because the row's `verticalAlignment` is `center`.

I also experimented with an API that allowed you to specify separate alignments for each of these tasks, but it just came out over-engineered and harder to understand.

> it's weird one outlier element can cause others' positions to be changed

This part does match SwiftUI, and it's key to making relative alignments work with values from the elements' own coordinate spaces. If we pushed outliers outside the stack bounds, we wouldn't be able to do things like baseline alignment, because everything would be getting pushed out.

Reply:
Got it, that makes sense.